src/tsig/mod.rs

```diff
@@ -1311,23 +1333,29 @@ impl<'a, Octs: Octets + ?Sized> MessageTsig<'a, Octs> {
     /// Checks that there is exactly one TSIG record in the additional
     /// section, that it is the last record in this section. If that is true,
     /// returns the parsed TSIG records.
-    fn from_message(msg: &'a Message<Octs>) -> Option<Self> {
-        let mut section = msg.additional().ok()?;
+    fn from_message(msg: &'a Message<Octs>) -> Result<Self, TsigError> {
+        let mut section = msg.additional().map_err(|_| TsigError::Missing)?;
+
+        // Find the first TSIG record, which we will assert to be the last
+        // one to verify that it is the only one.
         let mut start = section.pos();
-        let mut record = section.next()?;
-        loop {
-            let record_start = section.pos();
-            record = match section.next() {
-                Some(record) => record,
-                None => break,
-            };
-            start = record_start;
+        while let Some(record) = section.next() {
+            let record = record
+                .map_err(|_| TsigError::Missing)?
+                .into_record()
+                .map_err(|_| TsigError::Invalid)?;
+
+            if let Some(record) = record {
+                // We got a valid TSIG, now assert that it's the last record:
+                if section.next().is_some() {
+                    return Err(TsigError::Position);
+                }
+                return Ok(MessageTsig { record, start });
+            }
+            start = section.pos();
```

Comment:
We don’t really need to mutable and updated `start` any more. You can just grab `section.pos()` at the start of the iteration now.